File: 1 - LinkedList/Kotlin/2-BasicDoubleLinkedList/doubleLinkedList.kt
// Node class representing each element in the doubly linked list
class Node(
    var name: String,     // Data title
    var username: String, // Author of the data
    var email: String,    // Email of the user
    var password: String  // Password of the user
) {
    var prev: Node? = null // Pointer to the previous node in the list
    var next: Node? = null // Pointer to the next node in the list
}

// LinkedList class to manage the doubly linked list
class LinkedList {
    private var head: Node? = null // Pointer to the head of the list
    private var tail: Node? = null // Pointer to the tail of the list
    private var size: Int = 0      // Size of the list

    // Function to create a doubly linked list with the first node
    fun createDoubleLinkedList(name: String, username: String, email: String, password: String) {
        val newNode = Node(name, username, email, password)
        head = newNode
        tail = newNode
        size = 1
    }

    // Function to add a node at the beginning of the list
    fun addAtFirstNode(name: String, username: String, email: String, password: String) {
        if (size == 0) {
            createDoubleLinkedList(name, username, email, password)
            return
        }
        val newNode = Node(name, username, email, password)
        newNode.next = head
        head?.prev = newNode
        head = newNode
        size++
    }

    // Function to add a node at the nth position in the list
    fun addAtNthNode(name: String, username: String, email: String, password: String, pos: Int) {
        if (pos < 1 || pos > size + 1) {
            println("Invalid position!")
            return
        }
        if (pos == 1) {
            addAtFirstNode(name, username, email, password)
            return
        }
        if (pos == size + 1) {
            addAtLastNode(name, username, email, password)
            return
        }
        val newNode = Node(name, username, email, password)
        var cur = head
        for (i in 1 until pos - 1) {
            cur = cur?.next
        }
        newNode.next = cur?.next
        newNode.prev = cur
        cur?.next?.prev = newNode
        cur?.next = newNode
        size++
    }

    // Function to add a node at the end of the list
    fun addAtLastNode(name: String, username: String, email: String, password: String) {
        if (size == 0) {
            createDoubleLinkedList(name, username, email, password)
            return
        }
        val newNode = Node(name, username, email, password)
        newNode.prev = tail
        tail?.next = newNode
        tail = newNode
        size++
    }

    // Function to update the first node in the list
    fun updateFirstNode(name: String, username: String, email: String, password: String) {
        if (size == 0) {
            println("List is empty!")
            return
        }
        head?.name = name
        head?.username = username
        head?.email = email
        head?.password = password
    }

    // Function to update the nth node in the list
    fun updateNthNode(name: String, username: String, email: String, password: String, pos: Int) {
        if (pos < 1 || pos > size) {
            println("Invalid position!")
            return
        }
        var cur = head
        for (i in 1 until pos) {
            cur = cur?.next
        }
        cur?.name = name
        cur?.username = username
        cur?.email = email
        cur?.password = password
    }

    // Function to update the last node in the list
    fun updateLastNode(name: String, username: String, email: String, password: String) {
        if (size == 0) {
            println("List is empty!")
            return
        }
        tail?.name = name
        tail?.username = username
        tail?.email = email
        tail?.password = password
    }

    // Function to remove the first node from the list
    fun removeAtFirst() {
        if (size == 0) {
            println("List is empty!")
            return
        }
        if (size == 1) {
            head = null
            tail = null
        } else {
            head = head?.next
            head?.prev = null
        }
        size--
    }

    // Function to remove the nth node from the list
    fun removeAtNthNode(pos: Int) {
        if (pos < 1 || pos > size) {
            println("Invalid position!")
            return
        }
        if (pos == 1) {
            removeAtFirst()
            return
        }
        if (pos == size) {
            removeAtLast()
            return
        }
        var cur = head
        for (i in 1 until pos) {
            cur = cur?.next
        }
        cur?.prev?.next = cur?.next
        cur?.next?.prev = cur?.prev
        size--
    }

    // Function to remove the last node from the list
    fun removeAtLast() {
        if (size == 0) {
            println("List is empty!")
            return
        }
        if (size == 1) {
            head = null
            tail = null
        } else {
            tail = tail?.prev
            tail?.next = null
        }
        size--
    }

    // Function to print the doubly linked list
    fun printDoubleLinkedList() {
        if (size == 0) {
            println("List is empty!")
            return
        }
        println("Number of Nodes: ${countNodes()}") // Display the total number of nodes
        var cur = head
        var i = 1
        while (cur != null) {
            println("Node $i")
            println("Name\t\t: ${cur.name}")
            println("Username\t: ${cur.username}")
            println("Email\t\t: ${cur.email}")
            println("Password\t: ${cur.password}\n")
            cur = cur.next
            i++
        }
    }

    // Utility function to count the number of nodes in the list
    fun countNodes(): Int {
        return size // Return the total number of nodes
    }
}

fun main() {
    val list = LinkedList()

    // Create the doubly linked list with the first node
    list.createDoubleLinkedList("Arief F-sa", "ariefff", "arief@mail.id", "Arief123")
    list.printDoubleLinkedList()
    println("==========\n\n\n")

    // Add a node at the end of the list
    list.addAtLastNode("Fadly Pejel", "pejeel", "fadly@mail.id", "Fadly123")
    list.printDoubleLinkedList()
    println("==========\n\n\n")

    // Add a node at the beginning of the list
    list.addAtFirstNode("Nisa Uswa", "icaaa", "nisa@mail.id", "Nisa123")
    list.printDoubleLinkedList()
    println("==========\n\n\n")

    // Remove the first node from the list
    list.removeAtFirst()
    list.printDoubleLinkedList()
    println("==========\n\n\n")

    // Remove the last node from the list
    list.removeAtLast()
    list.printDoubleLinkedList()
    println("==========\n\n\n")

    // Add multiple nodes at the beginning of the list
    list.addAtFirstNode("dummyE", "dummy_5", "dummy5@mail.id", "Dummy123")
    list.addAtFirstNode("dummyD", "dummy_4", "dummy4@mail.id", "Dummy123")
    list.addAtFirstNode("dummyC", "dummy_3", "dummy3@mail.id", "Dummy123")
    list.addAtFirstNode("dummyB", "dummy_2", "dummy2@mail.id", "Dummy123")
    list.addAtFirstNode("dummyA", "dummy_1", "dummy1@mail.id", "Dummy123")
    list.printDoubleLinkedList()
    println("==========\n\n\n")

    // Add a node at the nth position in the list
    list.addAtNthNode("Ayang", "icaaa", "nisa@mail.id", "Nisa123", 5)
    list.printDoubleLinkedList()
    println("==========\n\n\n")

    // Remove the nth node from the list
    list.removeAtNthNode(6)
    list.printDoubleLinkedList()
    println("==========\n\n\n")
}

File: 1 - LinkedList/Go/2-BasicDoubleLinkedList/doubleLinkedList.go
package main

import (
    "fmt"
)

// Node struct representing each element in the doubly linked list
type Node struct {
    Name     string // Data title
    Username string // Author of the data
    Email    string // Author of the data
    Password string // Publication year of the data
    Prev     *Node  // Pointer to the previous data in the list
    Next     *Node  // Pointer to the next data in the list
}

// LinkedList struct to manage the doubly linked list
type LinkedList struct {
    Head *Node // Pointer to the head of the list
    Tail *Node // Pointer to the tail of the list
    Size int   // Size of the list
}

// NewLinkedList function to initialize an empty list
func NewLinkedList() *LinkedList {
    return &LinkedList{nil, nil, 0}
}

// CreateDoubleLinkedList function to create a doubly linked list with the first node
func (list *LinkedList) CreateDoubleLinkedList(name, username, email, password string) {
    newNode := &Node{name, username, email, password, nil, nil}
    list.Head = newNode
    list.Tail = newNode
    list.Size = 1
}

// AddAtFirstNode function to add a node at the beginning of the list
func (list *LinkedList) AddAtFirstNode(name, username, email, password string) {
    if list.Size == 0 {
        list.CreateDoubleLinkedList(name, username, email, password)
        return
    }
    newNode := &Node{name, username, email, password, nil, list.Head}
    list.Head.Prev = newNode
    list.Head = newNode
    list.Size++
}

// AddAtNthNode function to add a node at the nth position in the list
func (list *LinkedList) AddAtNthNode(name, username, email, password string, pos int) {
    if pos < 1 || pos > list.Size+1 {
        fmt.Println("Invalid position!")
        return
    }
    if pos == 1 {
        list.AddAtFirstNode(name, username, email, password)
        return
    }
    if pos == list.Size+1 {
        list.AddAtLastNode(name, username, email, password)
        return
    }
    newNode := &Node{name, username, email, password, nil, nil}
    cur := list.Head
    for i := 1; i < pos-1; i++ {
        cur = cur.Next
    }
    newNode.Next = cur.Next
    newNode.Prev = cur
    cur.Next.Prev = newNode
    cur.Next = newNode
    list.Size++
}

// AddAtLastNode function to add a node at the end of the list
func (list *LinkedList) AddAtLastNode(name, username, email, password string) {
    if list.Size == 0 {
        list.CreateDoubleLinkedList(name, username, email, password)
        return
    }
    newNode := &Node{name, username, email, password, list.Tail, nil}
    list.Tail.Next = newNode
    list.Tail = newNode
    list.Size++
}

// UpdateFirstNode function to update the first node in the list
func (list *LinkedList) UpdateFirstNode(name, username, email, password string) {
    if list.Size == 0 {
        fmt.Println("List is empty!")
        return
    }
    list.Head.Name = name
    list.Head.Username = username
    list.Head.Email = email
    list.Head.Password = password
}

// UpdateNthNode function to update the nth node in the list
func (list *LinkedList) UpdateNthNode(name, username, email, password string, pos int) {
    if pos < 1 || pos > list.Size {
        fmt.Println("Invalid position!")
        return
    }
    cur := list.Head
    for i := 1; i < pos; i++ {
        cur = cur.Next
    }
    cur.Name = name
    cur.Username = username
    cur.Email = email
    cur.Password = password
}

// UpdateLastNode function to update the last node in the list
func (list *LinkedList) UpdateLastNode(name, username, email, password string) {
    if list.Size == 0 {
        fmt.Println("List is empty!")
        return
    }
    list.Tail.Name = name
    list.Tail.Username = username
    list.Tail.Email = email
    list.Tail.Password = password
}

// RemoveAtFirst function to remove the first node from the list
func (list *LinkedList) RemoveAtFirst() {
    if list.Size == 0 {
        fmt.Println("List is empty!")
        return
    }
    if list.Size == 1 {
        list.Head = nil
        list.Tail = nil
    } else {
        list.Head = list.Head.Next
        list.Head.Prev = nil
    }
    list.Size--
}

// RemoveAtNthNode function to remove the nth node from the list
func (list *LinkedList) RemoveAtNthNode(pos int) {
    if pos < 1 || pos > list.Size {
        fmt.Println("Invalid position!")
        return
    }
    if pos == 1 {
        list.RemoveAtFirst()
        return
    }
    if pos == list.Size {
        list.RemoveAtLast()
        return
    }
    cur := list.Head
    for i := 1; i < pos; i++ {
        cur = cur.Next
    }
    cur.Prev.Next = cur.Next
    cur.Next.Prev = cur.Prev
    list.Size--
}

// RemoveAtLast function to remove the last node from the list
func (list *LinkedList) RemoveAtLast() {
    if list.Size == 0 {
        fmt.Println("List is empty!")
        return
    }
    if list.Size == 1 {
        list.Head = nil
        list.Tail = nil
    } else {
        list.Tail = list.Tail.Prev
        list.Tail.Next = nil
    }
    list.Size--
}

// PrintDoubleLinkedList function to print the doubly linked list
func (list *LinkedList) PrintDoubleLinkedList() {
    if list.Size == 0 {
        fmt.Println("List is empty!")
        return
    }
    fmt.Printf("Number of Nodes: %d\n", list.CountNodes()) // Display the total number of nodes
    cur := list.Head
    i := 1
    for cur != nil {
        fmt.Printf("Node %d\n", i)
        fmt.Printf("Name\t\t: %s\n", cur.Name)
        fmt.Printf("Username\t: %s\n", cur.Username)
        fmt.Printf("Email\t\t: %s\n", cur.Email)
        fmt.Printf("Password\t: %s\n\n", cur.Password)
        cur = cur.Next
        i++
    }
}

// CountNodes utility function to count the number of nodes in the list
func (list *LinkedList) CountNodes() int {
    return list.Size // Return the total number of nodes
}

func main() {
    list := NewLinkedList()

    // Create the doubly linked list with the first node
    list.CreateDoubleLinkedList("Arief F-sa", "ariefff", "arief@mail.id", "Arief123")
    list.PrintDoubleLinkedList()
    fmt.Println("==========\n\n\n")

    // Add a node at the end of the list
    list.AddAtLastNode("Fadly Pejel", "pejeel", "fadly@mail.id", "Fadly123")
    list.PrintDoubleLinkedList()
    fmt.Println("==========\n\n\n")

    // Add a node at the beginning of the list
    list.AddAtFirstNode("Nisa Uswa", "icaaa", "nisa@mail.id", "Nisa123")
    list.PrintDoubleLinkedList()
    fmt.Println("==========\n\n\n")

    // Remove the first node from the list
    list.RemoveAtFirst()
    list.PrintDoubleLinkedList()
    fmt.Println("==========\n\n\n")

    // Remove the last node from the list
    list.RemoveAtLast()
    list.PrintDoubleLinkedList()
    fmt.Println("==========\n\n\n")

    // Add multiple nodes at the beginning of the list
    list.AddAtFirstNode("dummyE", "dummy_5", "dummy5@mail.id", "Dummy123")
    list.AddAtFirstNode("dummyD", "dummy_4", "dummy4@mail.id", "Dummy123")
    list.AddAtFirstNode("dummyC", "dummy_3", "dummy3@mail.id", "Dummy123")
    list.AddAtFirstNode("dummyB", "dummy_2", "dummy2@mail.id", "Dummy123")
    list.AddAtFirstNode("dummyA", "dummy_1", "dummy1@mail.id", "Dummy123")
    list.PrintDoubleLinkedList()
    fmt.Println("==========\n\n\n")

    // Add a node at the nth position in the list
    list.AddAtNthNode("Ayang", "icaaa", "nisa@mail.id", "Nisa123", 5)
    list.PrintDoubleLinkedList()
    fmt.Println("==========\n\n\n")

    // Remove the nth node from the list
    list.RemoveAtNthNode(6)
    list.PrintDoubleLinkedList()
    fmt.Println("==========\n\n\n")
}